### ISSUE TITLE
perf: matmul accumulation does not need to default to float

### ISSUE THREAD

**Issue by jon-chuang** [Dao-AILab/flash-attention#543]
As we can see, it is hardcoded to float here:
https://github.com/Dao-AILab/flash-attention/blob/d0032700d1c7a1353a3a8f2fadfbc73b2dc6b5dc/csrc/flash_attn/src/kernel_traits.h#L26

However, it is known that fp16 accumulation is much faster for matmuls (see e.g. https://pytorch.org/docs/stable/notes/cuda.html#reduced-precision-reduction-in-fp16-gemms).

**Comment by tridao** [comment]
Accumulation is done in fp32 for numerical precision and numerical stability. Fp16 might get inf if e.g. Q @ K^T has entries larger than 65k.

**Comment by jon-chuang** [comment]
Do you think it may be feasible to use `fp16` accumulation for `P @ V`? We know that `P` is normalized to 1.

**Comment by jon-chuang** [comment]
Notes:


### Handling of delayed softmax reciprocal case:

Enforce numerical stability for non-inf values of V: `max(O) <= max(V) <= max(fp16)`:
 
If we choose a factor `stability_factor = ceil(log(seq_len))`:

Then we know
`max(O_ij) = max(sum_{k<j} (exp(S-amax-stability_factor) @ V_ij)) < 1/seq_len * (j + 1) * block_k * max(V_i0, ..., V_ij) <=  max(V_i0, ..., V_ij) <= max(V)`

So we simply need to exponentiate: `S - amax - stability_factor`. Of course, we may get poor numerical accuracy by squashing towards to 0.

---

Due to uncertain numerical stability for delayed reciprocal, it seems wise to apply the fp16 accumulation only to the in-loop reciprocal case.

I believe that one should benchmark both cases (`fp16 acc + in-loop`, `fp32 acc + delayed reciprocal`).

**Comment by tridao** [comment]
Interesting idea!
There are 2 related things here:
- Numerical range (stability): do we have enough range to avoid Inf, NaN. Maybe with the stability factor you mentioned, or just focusing on P @ V could work.
- Numerical precision: how close is the output to the standard attention in fp32 (reference). Because of the softmax rescaling in FlashAttention where we scale the output of each block by sth potentially very small, I've always kept all the scaling in fp32. As a result, FlashAttention is usually closer to fp32 reference (i.e. more accurate) than even a standard attention implementation in fp16/bf16.

I'm not sure how much faster fp16 accumulate could be. Yeah we should benchmark. Maybe it's easier to do that with the Triton implementation?

**Comment by jon-chuang** [comment]
Benchmarks [here](https://github.com/google/jax/pull/17328#issuecomment-1720262896)

Have yet to explore numerical accuracy (I think in the first place numerical accuracy may be lower for Triton/Pallas)

**Comment by tridao** [comment]
For consumer cards (e.g., 4090) matmul with fp16 accum has twice the throughput of matmul with fp32 accum. Maybe that explains some of the speed difference you're seeing. However, for data center cards (e.g., A100) I think the throughput is the same. The benefit there is probably fewer registers needed to hold the accumulators.

**Comment by jon-chuang** [comment]
Yes, interestingly enough and rather sadly:

Ada Lovelace:
![image](https://github.com/Dao-AILab/flash-attention/assets/9093549/75ca88e8-004b-417c-b120-afbe8583983a)

A100/H100:
![image](https://github.com/Dao-AILab/flash-attention/assets/9093549/5cfdb23d-4143-46c9-bdf4-43d99b4fb270)

**Comment by jon-chuang** [comment]
After some further thought, I think for H100, it is worth investigating using more FP16 arithmetic. That's because FP32 arithmetic on H100 is _unbelievably expensive_ (16.7x), with FP32 TFLOPs _lower than RTX 4090_ (60 v.s. 82.6) in contrast with FP16 TFLOPs (120 v.s. 82.6).

I don't think _tensor core_ is the bottleneck for H100 perf. 

While it might be that memory bandwidth is bottleneck, I believe that investigating using FP16 arithmetic - or otherwise further reducing non-matmul arithmetic - in the main loop will be extremely productive.

![image](https://github.com/Dao-AILab/flash-attention/assets/9093549/5384fa00-fbf9-47a5-9bdb-01bcbe4b5b56)

**Comment by jon-chuang** [comment]
An interesting datapoint from [Adept.ai Persimmon 8B](https://www.adept.ai/blog/persimmon-8b):

> We add layernorm to the Q and K embeddings before they enter the attention calculation.

This should help `Q @ K` float accuracy. So I believe that even `Q @ K` fp16 accumulation stability can be feasible with the right model choice.

**Comment by Zhao-Dongyu** [comment]
Recently, I conducted some related experiments inspired by your discussion two years ago. @tridao @jon-chuang 

First, I took a look at the performance of modifying GEMM to use FP16 accumulation. 

<img width="830" height="483" alt="Image" src="https://github.com/user-attachments/assets/da935edd-8a76-4db3-b9b9-8a975ce59c22" />

<img width="835" height="501" alt="Image" src="https://github.com/user-attachments/assets/67064064-a344-4fff-8f45-72632a9606b1" />

After modifying the source code of flash-attn and evaluating its performance on a 4090 card, I found that the throughput almost doubled, which is very impressive!

| |Accumulate|headdim|causal|seqlen|TFLOPS|
|---|:---:|:---:|:---:|:---:|:---:|
|Fav2|FP32|128|False|8192|137.9|
|Fav3|FP32|128|False|8192|142.2|
|Fav3|FP16|128|False|8192|**261.4**|
|Fav2|FP32|128|True|8192|136.2|
|Fav3|FP32|128|True|8192|144.9|
|Fav3|FP16|128|True|8192|**254.6**|

Then, I analyzed the data distribution of the intermediate results of $Q · K^T$ and the attention results in the 28 layers of the Qwen2.5 model. I found that there are indeed cases where the values exceed 65504. Therefore, it is crucial to clamp the infinities; otherwise, the model output will be abnormal.

<img width="5971" height="5366" alt="Image" src="https://github.com/user-attachments/assets/f67d3a3a-311b-410e-b519-1652aab13533" />

Finally, I modified the linear layers in vllm and achieved approximately a 15% speedup end-to-end. After further modifying flash-attn, I obtained an additional 31% speedup. 

**The overall end-to-end speedup reached around 50%, which is quite significant.** 

It should be noted that there is indeed some precision loss, but **the initial evaluation indicates that it is acceptable.**
